R/parse-body.R

```diff
@@ -192,54 +243,56 @@ parser_multi <- function() {
     toparse <- parse_multipart(value, boundary)
     # content-type detection
     lapply(toparse, function(x) {
-      if (!is.null(x$filename)) {
-        x$content_type <- getContentType(tools::file_ext(x$filename))
+      if (
+        is.null(x$content_type) ||
+        # allows for files to be shipped as octect, but parsed using the matching value in `knownContentTypes`
+        # (Ex: `.rds` files -> `application/rds` which has a proper RDS parser)
+        isTRUE(stri_detect_fixed(x$content_type, "application/octet-stream"))
+      ) {
+        if (!is.null(x$filename)) {
+          # Guess content-type from file extension
+          x$content_type <- getContentType(tools::file_ext(x$filename))
+        }
       }
-      parseRaw(x)
+      parse_raw(x)
     })
   }
 }
 
 
 
+add_parsers_onLoad <- function() {
 
-#' OCTET
-#' @rdname parsers
-#' @export
-parser_octet <- function() {
-  function(value, filename = NULL, ...) {
-    attr(value, "filename") <- filename
-    return(value)
-  }
-}
+  # add both `application/XYZ` and `text/XYZ` parsers
+  for (type in c("application", "text")) {
+    mime_type <- function(x) {
+      paste0(type, "/", x)
+    }
 
+    add_parser(mime_type("json"),                 parser_json())
 
+    add_parser(mime_type("csv"),                  parser_csv())
+    add_parser(mime_type("x-csv"),                parser_csv())
 
+    add_parser(mime_type("yaml"),                 parser_yaml())
+    add_parser(mime_type("x-yaml"),               parser_yaml())
+
+    add_parser(mime_type("tab-separated-values"), parser_tsv())
 
-#' YAML
-#' @rdname parsers
-#' @export
-parser_yaml <- function() {
-  if (!requireNamespace("yaml", quietly = TRUE)) {
-    stop("yaml must be installed for the yaml parser to work")
-  }
-  function(value, content_type = NULL, ...) {
-    charset <- getCharacterSet(content_type)
-    value <- rawToChar(value)
-    Encoding(value) <- charset
-    yaml::yaml.load(value)
   }
-}
 
+  # only one form of these parsers
+  add_parser("application/x-www-form-urlencoded", parser_query())
+  add_parser("application/rds",                   parser_rds())
+  add_parser("multipart/form-data",               parser_multi())
+  add_parser("application/octet",                 parser_octet())
 
+  add_parser("text/plain", parser_text())
```

Comment:
shoud we add text/html? It will eventually be picked up by the picker in the next if check anyway?

Reply:
My thought is to leave it open as it isn't a _solution_ to `text/html`.

I'd expect `text/html` or `application/html` to turn into tag objects via something like https://rvest.tidyverse.org/reference/html.html . I'd like to not make that design decision for now.

Reply:
Makes perfect sense.